### PR TITLE
Use narwhals for dataframe support

### LINKED ISSUE #210: Use narwhals for dataframe support
As this is a new project, I suggest we work with https://github.com/narwhals-dev/narwhals so that we have API-agnostic DataFrame support. In particular, Polars for large data sets.

### PR DESCRIPTION
**Pull request by juanitorduz**
Closes https://github.com/pymc-labs/pathmc/issues/210

### CODE REVIEW

**Comment by juanitorduz** [comment]
@bugbot review

File: pyproject.toml [hunk 1]
```diff
@@ -32,9 +32,9 @@ classifiers = [
 dependencies = [
     "arviz<1",
     "graphviz",
+    "narwhals>=1.20",
     "networkx",
     "numpy<2.4",
-    "pandas",
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pyproject.toml:37]: fixed in https://github.com/pymc-labs/pathmc/pull/215/commits/c30eaddb281ad6c8c77c3fd2b8012f92961c05d1

**Comment by juanitorduz** [comment]
@bugbot review

File: pathmc/compile.py [hunk 2]
```diff
@@ -408,8 +415,8 @@ def compile_to_pymc(
     sparse_data: dict[str, np.ma.MaskedArray] = {}
     for reg in spec.regressions:
         v = reg.lhs
-        if v not in latent and v in data.columns and data[v].isna().any():
-            vals = np.asarray(data[v].values, dtype=float)
+        if v not in latent and v in data.columns and data[v].is_null().any():
```
[the remark this answers is not in the document]
Comment (juanitorduz) [reply pathmc/compile.py]: fixed in https://github.com/pymc-labs/pathmc/pull/215/commits/a2c800120708d44342af8bfc8e57b7edb8462b9f

**Comment by juanitorduz** [comment]
@bugbot review

**Comment by daimon-pymclabs** [comment]
## Code Review — narwhals dataframe support

### Overview

Solid implementation of the narwhals pattern. The boundary conversion approach (`nw.from_native` on entry, `.to_native()` on exit) is correct, and the internal code consistently uses `nw.DataFrame` throughout. Close to merge-ready with a few things worth addressing.

---

### Strengths

- `simulate()` correctly uses `IntoFrameT` for same-backend passthrough; `model()` uses `IntoFrame` for input-only — right call in both cases.
- `_partial_correlation_test` in `identify.py` replaces `dropna()` with numpy-space NaN filtering; the inline comment explaining the null/NaN unification across backends is genuinely useful.
- `add_lags` replacement of `groupby + shift` with `nw.col(var).shift(k).over(unit_col)` is idiomatic and cleaner than the pandas version.

---

### Issues / Suggestions

**1. `effects.py` and `identify.py` function bodies not fully updated?**

The signatures now accept `nw.DataFrame`, but if the function bodies still call pandas-specific methods (`.isna()`, `.values`, `.iloc`, `.loc`, `.reset_index()`, etc.), polars inputs will fail at runtime. Recommend grepping both files for those patterns to confirm they're clean.

**2. Patsy forces pandas conversion for polars users**

```python
dm = patsy.dmatrix(formula_str, data=data.to_pandas(), return_type="dataframe")
```

This is unavoidable since patsy only understands pandas, but worth a docstring note like *"polars inputs are converted to pandas internally for the patsy call"* so future contributors don't try to remove the `.to_pandas()`.

**3. Empty design matrix has no dtype**

```python
nw.from_dict({c: [] for c in cols}, backend=data.implementation)
```

Empty lists give narwhals no type information. Polars will infer `Null` dtype, which could surface surprising errors downstream. Safer:

```python
nw.from_dict({c: np.array([], dtype=float) for c in cols}, backend=data.implementation)
```

**4. `data[var].mean()` is not None-safe**

In `simulate.py` (two places):

```python
fill = float(data[var].mean())
```

narwhals `Series.mean()` returns `float | None` when the series is all-null. `float(None)` raises `TypeError`. The old pandas path returned `nan` silently. Fix:

```python
fill = float(data[var].mean() or np.nan)
```

**5. `with_columns(**new_columns)` — keyword dict unpacking**

```python
result = nw_data.with_columns(**new_columns)
```

`new_columns` is `dict[str, nw.Series]`. narwhals `with_columns` doesn't consistently support `**kwargs` with Series values across all versions. Safer form:

```python
result = nw_data.with_columns(list(new_columns.values()))
```

**6. Generator expression passed to `with_columns`**

```python
data_sim = data_sim.with_columns(nw.lit(0.0).alias(var) for var in zero_cols)
```

Passing a generator to `with_columns(*exprs)` works today but is fragile. Prefer an explicit list: `[nw.lit(0.0).alias(var) for var in zero_cols]`.

---

### Test Coverage Gaps

- **Panel models with polars** — no test for `model(..., panel={...})` with a polars input. This is the code path with the most bespoke narwhals translation (`with_row_index`, `sort`, `.over()`).
- **`test_implications` / identification** — signature was updated but not covered by the new tests.
- **Counterfactuals** — `att()`, `atu()`, `prob()` all access `self._data` but aren't tested with polars.
- **Tolerance in `test_fit_recovers_parameters_consistently`** — `atol=1e-10` / `rtol=1e-10` is very tight for MCMC-derived quantities. `rtol=1e-6` would be more robust against any future backend-path divergence.

---

### Minor: narwhals version bound

`narwhals>=1.20` is very loose. The lock pins `2.22.1` so it's fine in practice, but with narwhals 2.x out it may be worth tightening to `>=1.20,<3` to avoid unexpected breakage when 3.0 arrives.

**Comment by juanitorduz** [comment]
Thanks for the thorough review! Addressed in bd89b33. Disposition per point:

**1. effects.py / identify.py bodies** — verified clean. Every `pd.*` / `.values` use is on arviz/xarray output objects or the output result tables (by design), not on the input frame: input reads use narwhals (`data[col].std()`, `data.select(cols).to_numpy()`, `set(data.columns)`). No input-frame change needed. (Separately, the `.values` on the xarray objects was switched to `.to_numpy()` for hygiene.)

**2. Patsy → pandas** — added a `Notes` section to `build_design_matrix`'s docstring explaining the `data.to_pandas()` round-trip and that it must not be removed for non-pandas inputs.

**3. Empty design dtype** — fixed: `{c: np.array([], dtype=float) for c in cols}` so polars infers `Float64` instead of `Null`.

**4. `mean()` None-safety** — fixed, but **not** with `float(x or np.nan)`: that turns a legitimate mean of `0.0` into NaN (`0.0 or np.nan == np.nan`). Used an explicit `None` check instead, and applied the same guard to the two `std()` reads in `effects.py` (a zero-variance column would hit the same trap).

**5. `with_columns(**new_columns)`** — switched to the positional `with_columns(list(new_columns.values()))` (the series are already named).

**6. Generator → list** — done.

**Minor (version bound)** — tightened to `narwhals>=1.20,<3`.

**Test gaps** — added polars coverage in `tests/test_narwhals_backends.py`:
- `TestPanelBothBackends` — panel `with_row_index`/`sort`/`.over()` path (compile for both backends + slow fit & `do(simulate_over="time")`).
- `TestIdentificationBothBackends` — `test_implications()` on polars input.
- `TestCounterfactualsBothBackends` — `att()`/`atu()`/`prob()` on a polars-fitted model.
- Loosened `test_fit_recovers_parameters_consistently` tolerance from `1e-10` to `1e-6`.

Verification: `make check_lint` clean; `test-fast` 412 passed; slow suite 139 passed with only the 4 pre-existing #217 failures (no new failures).

**Comment by juanitorduz** [comment]
@bugbot review

**Comment by daimon-pymclabs** [comment]
## Round 2 Review

All 6 issues from the first review have been addressed. Here's the status:

### What was fixed ✅

- `effects.py`/`simulate.py` bodies audited: xarray `.values` → `.to_numpy()` throughout, including `extract_labeled_draws` and `compute_path_effect`.
- Empty design matrix now uses `np.array([], dtype=float)` so polars infers `Float64` not `Null`.
- `data[var].mean()` None-safety: both instances in `simulate.py` now use the explicit `if _mean is not None else float("nan")` guard (correctly avoids the `0.0 or nan` footgun).
- `with_columns(list(new_columns.values()))` — positional list, not `**kwargs`.
- narwhals bound tightened to `>=1.20,<3`.
- Tests expanded to 234 lines: `TestPanelBothBackends`, `TestIdentificationBothBackends`, `TestCounterfactualsBothBackends` all added. `rtol=1e-6` in consistency test.
- pandas explicitly re-declared as a core dependency.

---

### One remaining issue

**Generator not converted to list in `add_lags` (`panel.py`)**

The commit message says "explicit list comprehension" was applied everywhere, but this one was missed:

```python
result = result.with_columns(
    nw.col(var).shift(k).over(unit_col).alias(f"{var}_lag{k}")
    for var in variables        # ← still a generator
    for k in lag_orders
)
```

Should be:

```python
result = result.with_columns([
    nw.col(var).shift(k).over(unit_col).alias(f"{var}_lag{k}")
    for var in variables
    for k in lag_orders
])
```

---

### Minor (non-blocking)

`"__pos__"` is used as a temporary column name in both `compile.py` and `transforms.py` when computing sort indices. If a user's data contains a column with that name, the sort step would silently clobber it. Something like `"__nw_row_pos__"` would be more collision-resistant, but this is an edge case.

---

### Verdict

One small fix needed (the generator in `add_lags`), then this is ready to merge.

**Comment by drbenvincent** [comment]
## Review — narwhals dataframe support

Reviewed the full diff, the current branch state, and both prior review rounds. The refactor is clean and the boundary pattern (`nw.from_native` on entry, `.to_native()` on exit, `nw.DataFrame` internally) is applied consistently. All earlier Bugbot/review items are resolved, including the `add_lags` generator fix that landed via #228. One real concern remains plus a few non-blocking nits.

### Major concern — att/atu subgroup fill has inconsistent NaN semantics (likely a regression)

In `simulate.py` (both `run_do_pymc` branches), the exogenous fill value differs by code path:

```python
if subgroup_indices is not None:
    fill = float(data[var].to_numpy()[subgroup_indices].mean())   # numpy mean: PROPAGATES NaN
else:
    _mean = data[var].mean()                                      # narwhals mean: SKIPS nulls/NaN
    fill = float(_mean) if _mean is not None else float("nan")
```

The non-subgroup branch uses narwhals `Series.mean()`, which skips nulls/NaN (matching the old pandas behavior). The subgroup branch uses numpy `.mean()` on the raw array, which propagates NaN. So the two branches disagree on missing-value handling for the same column, and the subgroup branch is a regression from the old `data[var].iloc[subgroup_indices].mean()` (pandas, `skipna=True`).

Concretely: `att()`/`atu()` on a model with a partially-missing exogenous covariate will now return `NaN` where it previously returned a number. Since pathmc explicitly supports missing data via the masked-array path, this isn't purely hypothetical. This is exactly the "polars null/NaN edge cases are newly exercised" risk the PR summary itself flagged. Suggested fix: use `np.nanmean(...)` for the subgroup slice (or slice in narwhals space and reuse the same `mean()`/None guard so both branches share one code path).

### Minor / non-blocking

- **`__pos__` temp-column collision** (`compile.py` `_compile_scan_panel`, `transforms.py` `_apply_pymc_panel`): if user data already has a `__pos__` column, the row-index step interferes with it. A more distinctive sentinel like `__nw_row_pos__` is cheap insurance.
- **Two independent sorts for `data_sorted` vs `sort_idx`** in the panel scan path: correctness relies on both sorts ordering ties identically for the reshape/`reverse_idx` round-trip. It's safe today because balanced panels have unique `(unit, time)` keys (validated upstream), so ordering is deterministic — but reusing a single sort for both would be more obviously correct and robust if that invariant ever loosens.
- **Misplaced `# type: ignore[attr-defined]`** in `compute_path_effect` (`effects.py`): it now sits on the `.flatten()` line of the chained expression rather than the `idata.posterior` access it was meant for. Lint passes, so this is cosmetic.
- **Test gap:** the new cross-backend suite is solid, but nothing exercises *missing values* across pandas vs polars — the highest-risk surface (the subgroup-mean issue above, the `_partial_correlation_test` numpy-NaN filter, and the masked-outcome path). A small parametrized test injecting a NaN into an exogenous covariate would have caught the att/atu regression.

### Recommendation

Close to merge-ready. I'd treat the att/atu subgroup `mean()` NaN inconsistency as a should-fix before merge (one-line change, ideally with a missing-data cross-backend test); the rest is optional polish.

**Comment by juanitorduz** [comment]
Thanks @drbenvincent — addressed in d29fd78.

**Major concern (att/atu subgroup NaN regression) — fixed.** Both `run_do_pymc` branches now route the exogenous fill through a single `_exogenous_fill` helper, so the subgroup-slice and full-column paths share one NaN policy: skip nulls/NaN (restoring the old pandas `skipna=True` behavior) and map an all-missing column to `nan`. Implemented with `np.nanmean` guarded by an explicit all-NaN/empty check so there's no `RuntimeWarning` on an all-missing slice. `att()`/`atu()` no longer collapse to `NaN` when a treated row has a missing covariate.

**Test gap — closed.** Added `TestMissingDataBothBackends` to `tests/test_narwhals_backends.py`, parametrized over pandas/polars. Since an exogenous root covariate can't carry NaN through fitting (only endogenous vars get the masked-array path), it fits on complete data, injects a NaN into the stored covariate for a treated row, and asserts `att`/`atu` stay finite. It fails on the pre-fix numpy `.mean()` and passes after.

**Non-blocking nits — all done:**
- `__pos__` → `__nw_row_pos__` in `compile.py` and `transforms.py`.
- `_compile_scan_panel` now derives `data_sorted` and `sort_idx` from a single sort (sort once with the row index attached, then drop the helper column), removing the two-independent-sorts dependency.
- Moved the misplaced `# type: ignore[attr-defined]` in `compute_path_effect` onto the `idata.posterior[...]` access.

Verification: `make lint` clean; the touched suites (`test_narwhals_backends`, `test_att_atu`, `test_do`, `test_panel_do`, `test_transforms_do`, `test_effects`) pass.

**Comment by daimon-pymclabs** [comment]
## Round 3 (Final) Review

Two new commits since round 2. Both are clean. Here's what changed and my verdict.

---

### What's new since round 2

**Commit: "Fix att/atu subgroup exogenous-fill NaN semantics"**

- **`_exogenous_fill` helper** (`simulate.py`) — replaces the duplicated `data[var].mean()` calls with a shared helper that uses `np.nanmean` and explicitly guards against empty/all-NaN slices without emitting a numpy `RuntimeWarning`. Both the subgroup-slice and full-column paths now share the same NaN-skip policy, which was the root of the `att`/`atu` regression. ✅
- **Single sort in `_compile_scan_panel`** — old code called `data.sort_values(...)` twice to get `data_sorted` and `sort_idx`. New code does one `with_row_index("__nw_row_pos__").sort(...)` and derives both from the same call — cleaner and more efficient. ✅
- **`__pos__` → `__nw_row_pos__`** in `compile.py` and `transforms.py` — addresses the column-name collision concern from round 2. ✅
- **`TestMissingDataBothBackends`** — regression test that fits on clean data, injects a NaN into a covariate, and verifies `att`/`atu` stay finite. Well-designed. ✅

**Commit: "Harden narwhals null/None edge cases flagged by Bugbot"**

- **`_build_unit_index`** — the dict-comprehension replacement for `pandas.Series.map` now wraps in `try/except KeyError` and raises a clear `ValueError` naming the null/NaN unit id and the fix. The old pandas path silently returned a NaN index that broke downstream; this is strictly better. ✅
- **`PathModel.do()` min/max guard** — `col_min`/`col_max` now checked for `None` before `float()` cast, mirroring the `mean()`/`std()` guards already in place. ✅

---

### Anything to flag?

One cosmetic note: `TestMissingDataBothBackends` mutates `model._data` directly (a private attribute) to inject the NaN. This is fine for a regression test and avoids complexity, just worth knowing it couples the test to the internal interface name. Not a blocker.

---

### Verdict

All issues from all three review rounds are resolved. No remaining concerns. **Ready to merge.**

**Comment by juanitorduz** [comment]
@bugbot review

**Review by drbenvincent — APPROVED** [review]
